src/stream/src/executor/expand.rs

```diff
@@ -0,0 +1,181 @@
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt::Debug;
+
+use futures::StreamExt;
+use futures_async_stream::try_stream;
+use risingwave_common::array::PrimitiveArray;
+use risingwave_common::catalog::{Field, Schema};
+use risingwave_common::types::DataType;
+
+use super::error::StreamExecutorError;
+use super::*;
+
+pub struct ExpandExecutor {
+    input: BoxedExecutor,
+    schema: Schema,
+    pk_indices: PkIndices,
+    expanded_keys: Vec<Vec<usize>>,
+}
+
+impl ExpandExecutor {
+    pub fn new(
+        input: Box<dyn Executor>,
+        pk_indices: PkIndices,
+        expanded_keys: Vec<Vec<usize>>,
+    ) -> Self {
+        let mut schema = input.schema().to_owned();
+        schema
+            .fields
+            .push(Field::with_name(DataType::Int64, "flag"));
+        Self {
+            input,
+            schema,
+            pk_indices,
+            expanded_keys,
+        }
+    }
+
+    #[try_stream(ok = Message, error = StreamExecutorError)]
+    async fn execute_inner(self) {
+        #[for_await]
+        for msg in self.input.execute() {
+            let msg = msg?;
+            match msg {
+                Message::Chunk(chunk) => {
+                    // TODO: handle dummy chunk.
+                    let chunk = chunk.compact()?;
+                    let (data_chunk, ops) = chunk.into_parts();
+                    let cardinality = data_chunk.cardinality();
+                    let (columns, _) = data_chunk.into_parts();
+
+                    let null_columns: Vec<Column> = columns
+                        .iter()
+                        .map(|column| {
+                            let array = column.array_ref();
+                            let mut builder = array.create_builder(cardinality)?;
+                            (0..cardinality).try_for_each(|_i| builder.append_null())?;
+                            Ok::<Column, StreamExecutorError>(Column::new(Arc::new(
+                                builder.finish()?,
+                            )))
+                        })
+                        .try_collect()?;
+
+                    for (i, keys) in self.expanded_keys.iter().enumerate() {
+                        let mut new_columns = null_columns.clone();
+                        for key in keys {
+                            new_columns[*key] = columns[*key].clone();
+                        }
+                        let flags = Column::from(PrimitiveArray::<i64>::from_slice(&vec![
+                            Some(i as i64);
+                            cardinality
+                        ])?);
+                        new_columns.push(flags);
```

Comment:
The parts that build `new_columns` from `columns` of input chunk seems reusable between batch and streaming. Could you extract them to a common utility?